bases/rsptx/book_server_api/routers/rslogging.py

```diff
@@ -361,50 +360,74 @@ async def updatelastpage(
         tz_offset = float(values.get("tz_offset", 0))
     else:
         tz_offset = 0
-
-    if practice_settings and practice_settings.flashcard_creation_method == 0:
-        await add_flashcard(request_data, lpd, user, tz_offset)
+        
+    if practice_settings:
+        if request_data.markingComplete:
+            if practice_settings.flashcard_creation_method == 0:
+                # self-paced flashcard creation based on marking a page as complete
+                rslogger.debug(f"self-paced flashcard creation based on marking a page as complete\n{request_data=}")
+                course = await fetch_course(user.course_name)
+                await potentially_change_flashcard(course.base_course, lpd["last_page_chapter"], lpd["last_page_subchapter"], user, tz_offset, add=True)
+
+        elif request_data.markingIncomplete:
+            if practice_settings.flashcard_creation_method == 0:
+                course = await fetch_course(user.course_name)
+                rslogger.debug(f"self-paced flashcard deletion based on marking a page as incomplete\n{request_data=}")          
+                await potentially_change_flashcard(course.base_course, lpd["last_page_chapter"], lpd["last_page_subchapter"], user, tz_offset, remove=True)
+        elif request_data.pageLoad and practice_settings.flashcard_creation_method == 3:
+            # self-paced flashcard creation based on loading a page
+            course = await fetch_course(user.course_name)
+            await potentially_change_flashcard(course.base_course, lpd["last_page_chapter"], lpd["last_page_subchapter"], user, tz_offset, add=True)
 
     return make_json_response(detail="Success")
 
 
-async def add_flashcard(
```

Comment:
I'm definitely confused by this.  potentially_change_flashcard is semantically very different from add_flashcard.  And I wrote add flashcard knowing that when one uses peer instruction we want to add questions from peer instruction that people miss to the practice suite.  And there is another researcher that I'm pretty sure is using add_flashcard.

Reply:
With the monorepo -- We should be refactoring functions like this to a new component.  something like components/rsptx/practice

Reply:
``poetry poly add component --name practice`` would create the component for you.

Reply:
`add_flashcard` doesn't show up anywhere else in the repo. Is that what you mean by another researcher using it?

I think it might help for me to explain the mystery of only one flashcard being added here rather than a new flashcard for each question. The logic, it turns out, is that we add one flashcard (UserTopicPractice instance) for each *topic*, not for each question. 

Some of the comments and var names in the code before created confusion about this; I tried to clean some of them up.

So if there are researchers who've written code to add questions from peer instruction to existing topics (i.e., chapter/subchapter combos), that code should *not* be adding a new UserTopicPractice. It should just associate the questions with the topics. The only subtlety might be that if they add questions to subchapters that have zero built-in questions from the book, then they might need to create one UserTopicPractice (for each student) for that topic, because the current logic doesn't create a UserTopicPractice for subchapters with no practice questions.

If there are people who you think might be working on this and misunderstanding how it's supposed to work, please put them in touch with me.

As for starting a new component, I agree that that would be a good idea. It would make it easier for more things to start being migrated to that component over time if we get it started now. I have a feeling, however, that there will be more to it in adding a new component than just that one command you've given me. At a minimum, updating documentation in a few places. I think if I'm going to do this, we should probably do it together in a pair programming session on zoom.

Reply:
You are probably right.  I can already think of a couple of edits to pyproject.toml files that will need to be done.  But I think its worth the effort.  Happy to set up a time to zoom.